### PR TITLE
[GCB-sass] Refactor the SassTask to not use Gulp.

### PR DESCRIPTION
**Pull request by iclanton**
This PR builds upon jareha's work to support both SASS and SCSS: https://github.com/Microsoft/web-build-tools/pull/852

It also refactors the task to no longer use Gulp.

### CODE REVIEW

File: core-build/gulp-core-build-sass/src/SassTask.ts [hunk 1]
```diff
@@ -1,50 +1,68 @@
...
+import * as postcss from 'postcss';
+import * as CleanCss from 'clean-css';
+import * as autoprefixer from 'autoprefixer';
+import * as cssModules from 'postcss-modules';
+import * as crypto from 'crypto';
 
 export interface ISassTaskConfig {
-  /** An optional parameter for text to include in the generated typescript file. */
+  /**
+   * An optional parameter for text to include in the generated typescript file.
```
Comment (pgonzal) [inline core-build/gulp-core-build-sass/src/SassTask.ts]: >typescript [](start = 64, length = 10)

TypeScript

File: core-build/gulp-core-build-sass/src/SassTask.ts [hunk 2]
```diff
@@ -76,218 +107,214 @@ export class SassTask extends GulpTask<ISassTaskConfig> {
...
       completeCallback('taskConfig.sassMatch must be defined');
       return;
     }
 
-    /* tslint:disable:typedef */
-    const autoprefixer = require('autoprefixer');
-    const cssModules = require('postcss-modules');
-    /* tslint:enable:typedef */
+    return this._globAll(...this.taskConfig.sassMatch).then((matches: string[]) => {
+      return Promise.all(matches.map((match) => this._processFile(match)));
```
Comment (pgonzal) [inline core-build/gulp-core-build-sass/src/SassTask.ts]: >Promise [](start = 13, length = 7)

I feel like we should either be using `Promise` or `completeCallback`, but not both.

IMO the API itself was designed wrong -- it should have had separate signatures for the two usage patterns.

**Comment by pgonzal** [comment]
@jareha FYI

File: core-build/gulp-core-build-sass/src/SassTask.ts [hunk 3]
```diff
@@ -76,218 +107,214 @@ export class SassTask extends GulpTask<ISassTaskConfig> {
...
+      );
+    }).then((result: nodeSass.Result) => {
+      const options: postcss.ProcessOptions = {
+        from: filePath
+      };
+      if (result.map && !this.buildConfig.production) {
+        options.map = {
+          prev: result.map.toString()
+        };
+      }
```
Comment (pgonzal) [inline core-build/gulp-core-build-sass/src/SassTask.ts]: Add some comments here.  I wasn't even sure what data type "map" was

File: core-build/gulp-core-build-sass/src/SassTask.ts [hunk 4]
```diff
@@ -76,218 +107,214 @@ export class SassTask extends GulpTask<ISassTaskConfig> {
...
-        sass.logError.call(this, error);
-        completeCallback('Errors found in sass file(s).');
-      }))
-      .pipe(postcss(postCSSPlugins))
-      .pipe(cleancss({
+      const plugins: postcss.AcceptedPlugin[] = [
+        ...this._postCSSPlugins,
+        ...(processAsModuleCss ? this._modulePostCssAdditionalPlugins : [])
+      ];
+      return postcss(plugins).process(result.css.toString(), options).then((postCssResult) => postCssResult);
```
Comment (pgonzal) [inline core-build/gulp-core-build-sass/src/SassTask.ts]: >then((postCssResult) => postCssResult); [](start = 70, length = 39)

maybe add a comment explaining why we need this.  (Is it adapting a proprietary Promise to have standard Promise typings?)

**Comment by jareha** [comment]
Awesome work, y’all! Thanks!

File: core-build/gulp-core-build-sass/src/sass.schema.json [hunk 5]
```diff
@@ -23,14 +23,14 @@
     },
 
     "sassMatch": {
-      "title": "Sass File Glob Pattern",
+      "title": "SASS File Glob Pattern",
```
Comment (jareha) [inline core-build/gulp-core-build-sass/src/sass.schema.json]: This should probably go back to “Sass”. My mistake! (Always thought it was capitalized, as it’s an acronym. Nope!)

File: core-build/gulp-core-build-sass/src/SassTask.ts [hunk 6]
```diff
@@ -76,218 +107,214 @@ export class SassTask extends GulpTask<ISassTaskConfig> {
...
+        exportClassNames = classNamesLines.join(EOL);
+      }
+
+      let lines: string[] = [];
+
+      lines.push(this.taskConfig.preamble || '');
+
+      if (cssOutputPathAbsolute) {
+        lines = lines.concat([
+          `require('./${path.basename(cssOutputPathAbsolute)}');`,
```
Comment (pgonzal) [inline core-build/gulp-core-build-sass/src/SassTask.ts]: >('./${path.basename(cssOutputPathAbsolute)}' [](start = 18, length = 44)

These strings are not escaped properly.  Should we just use JSON.stringify()?

**Pull request merged**